torch/fx/passes/utils/matcher_utils.py

```diff
@@ -0,0 +1,230 @@
+from dataclasses import dataclass, field
+from collections import defaultdict
+import copy
+from torch.fx.graph import Graph
+from torch.fx.node import Node
+from torch.fx._compatibility import compatibility
+from typing import Dict, List, Set
+
+__all__ = ['SubgraphMatcher']
+
+
+@compatibility(is_backward_compatible=False)
+@dataclass
+class _InternalMatch():
+    # Node from which the match was found
+    anchor: Node
+    # Maps nodes in the pattern subgraph to nodes in the larger graph
+    nodes_map: Dict[Node, Node] = field(default_factory=dict)
+
+    # nodes in target graph that are matched placeholder in pattern
+    placeholder_nodes: List[Node] = field(default_factory=list)
+
+    # sink nodes in matched subgraph returned by output
+    returning_nodes: List[Node] = field(default_factory=list)
+
+    def __copy__(self):
+        return _InternalMatch(anchor=self.anchor, nodes_map=self.nodes_map.copy(),
+                              placeholder_nodes=self.placeholder_nodes.copy(),
+                              returning_nodes=self.returning_nodes.copy())
+
+@compatibility(is_backward_compatible=False)
+class SubgraphMatcher:
+    def __init__(self, pattern: Graph,
+                 match_output: bool = False,
+                 match_placeholder: bool = False,
+                 fully_contained: bool = True,
+                 remove_overlapping_matches: bool = True) -> None:
```

Comment:
That's a lot of bool configuration flags. Can we simplify?

`fully_contained` is technically unnecessary as `fully_contained = False` can be emulated by synthetically adding every intermediate node into the output node (which when `fully_contained == True` appears to denote "this node is permitted to be referenced outside of the subgraph.)

`match_output` and `match_placeholder` seem to have arisen because you are punning output/placeholder nodes to mean can be referenced outside of subgraph / pattern variable. What about having an explicit pattern variable node instead? (It would probably be a `call_function` node but you can give it a special reserved op function.)

Reply:
Removed fully_contained flag.

Reply:
you are right about punning output and placeholder nodes.  I guess in most of the cases, user won't need to explicitly match placeholder and output, so they won't need to worry about setting this flag. 
It' a bit less natural to explicitly create a call_function("output") node to match with the actually output. 
But If you feel strong about this, I can fix this in the upcoming PR, where I introduced the wildcard nodes.